app/bundles/CampaignBundle/Controller/ImportController.php

```diff
@@ -0,0 +1,280 @@
+<?php
+
+namespace Mautic\CampaignBundle\Controller;
+
+use Doctrine\Persistence\ManagerRegistry;
+use Mautic\CampaignBundle\Form\Type\CampaignImportType;
+use Mautic\CoreBundle\Controller\FormController;
+use Mautic\CoreBundle\Factory\MauticFactory;
+use Mautic\CoreBundle\Factory\ModelFactory;
+use Mautic\CoreBundle\Helper\CoreParametersHelper;
+use Mautic\CoreBundle\Helper\DateTimeHelper;
+use Mautic\CoreBundle\Helper\PathsHelper;
+use Mautic\CoreBundle\Helper\UserHelper;
+use Mautic\CoreBundle\Security\Permissions\CorePermissions;
+use Mautic\CoreBundle\Service\FlashBag;
+use Mautic\CoreBundle\Translation\Translator;
+use Mautic\FormBundle\Helper\FormFieldHelper;
+use Mautic\LeadBundle\Helper\Progress;
+use Psr\Log\LoggerInterface;
+use Psr\Log\LogLevel;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+use Symfony\Component\Filesystem\Filesystem;
+use Symfony\Component\Form\FormError;
+use Symfony\Component\Form\FormFactoryInterface;
+use Symfony\Component\HttpFoundation\File\Exception\FileException;
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpFoundation\Response;
+
+class ImportController extends FormController
+{
+    // Steps of the import
+    public const STEP_UPLOAD_ZIP      = 1;
+
+    public const STEP_PROGRESS_BAR    = 2;
+
+    public const STEP_IMPORT_FROM_ZIP = 3;
+
+    public function __construct(
+        FormFactoryInterface $formFactory,
+        FormFieldHelper $fieldHelper,
+        CoreParametersHelper $coreParametersHelper,
+        ManagerRegistry $doctrine,
+        MauticFactory $factory,
+        ModelFactory $modelFactory,
+        UserHelper $userHelper,
+        EventDispatcherInterface $dispatcher,
+        Translator $translator,
+        FlashBag $flashBag,
+        private RequestStack $requestStack,
+        CorePermissions $security,
+        private LoggerInterface $logger,
+        private PathsHelper $pathsHelper,
+    ) {
+        parent::__construct($formFactory, $fieldHelper, $doctrine, $factory, $modelFactory, $userHelper, $coreParametersHelper, $dispatcher, $translator, $flashBag, $requestStack, $security);
+    }
+
+    public function newAction(Request $request, $ignorePost = false): Response
+    {
+        if (!$this->security->isAdmin() && !$this->security->isGranted('campaign:imports:create')) {
+            return $this->accessDenied();
+        }
+
+        $forceStop = $request->get('cancel', false);
+        $step      = ($forceStop) ? self::STEP_UPLOAD_ZIP : $this->requestStack->getSession()->get('mautic.campaign.import.step', self::STEP_UPLOAD_ZIP);
```

Comment:
Do you think we could create 3 distinct actions (routes) for these steps instead of having this very long method? I know it's done this way in the CSV import but it's very hard to maintain and debug.